pkg/fuzzer: TestFuzz strikes back

Happened on 1f72c409b75ea2adfb5410615b8fffd7f91b2690 on the `ci-qemu` instance:

```
--- FAIL: TestFuzz (6.81s)
    testutil.go:35: seed=1749818855861212489
    fuzzer_test.go:183: executor output:
        connected to manager: procs=4 cover_edges=1 kernel_64_bit=1 slowdown=1 syscall_timeout=50 program_timeout=5000 features=0x10
        received info reply: covfilter=0
        recv exec request 1: type=0 flags=0x3 env=0x1c09 exec=0x10 size=14
        proc 0: handshaking to execute request 1
        proc 0: got handshake reply
        proc 0: start executing request 1
        proc 0: got output: [18ms] exec opts: reqid=1 type=0 procid=0 threaded=1 cover=0 comps=0 dedup=0 signal=0  sandbox=1/0/0/0 timeouts=50/5000/1 kernel_64_bit=1
        spawned worker pid 181239
        proc 0: got output: Fatal glibc error: malloc assertion failure in sysmalloc: (old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)
        proc 0: got output: Fatal glibc error: malloc assertion failure in sysmalloc: (old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)
        proc 0: got execute reply
        handle completion: completed=0 output_size=262144
        proc 0: output pipe EOF
        proc 0: restarting subprocess, current state 2 attempts 0
        proc 0: subprocess exit status 6
        SYZFAIL: failed to recv rpc
        fd=3 want=4 recv=0 n=0 (errno 9: Bad file descriptor)
    fuzzer_test.go:184: execution of simple program fails: call 0 failed with errno 998
FAIL
FAIL	github.com/google/syzkaller/pkg/fuzzer	10.342s
```

Context:
* https://github.com/google/syzkaller/issues/5674
* https://github.com/google/syzkaller/issues/5770

--- a/pkg/rpcserver/local.go
+++ b/pkg/rpcserver/local.go
@@ -17,7 +17,6 @@ import (
 	"github.com/google/syzkaller/pkg/signal"
 	"github.com/google/syzkaller/pkg/vminfo"
 	"github.com/google/syzkaller/prog"
-	"golang.org/x/sync/errgroup"
 )
 
 type LocalConfig struct {
@@ -43,15 +42,27 @@ func RunLocal(ctx context.Context, cfg *LocalConfig) error {
 		return err
 	}
 	defer localCtx.serv.Close()
-	// groupCtx will be cancelled once any goroutine returns an error.
-	eg, groupCtx := errgroup.WithContext(ctx)
-	eg.Go(func() error {
-		return localCtx.RunInstance(groupCtx, 0)
-	})
-	eg.Go(func() error {
-		return localCtx.serv.Serve(groupCtx)
-	})
-	return eg.Wait()
+
+	// Note: we must not stop the RPC server before we finish RunInstance.
+	// Otherwise, RPC server will close the connection, and executor may SYZFAIL
+	// on the closed network connection.
+	// We first need to wait for the executor binary to finish, and only then stop the RPC server.
+	// However, we want to stop both if the other one errors out.
+	instCtx, instCancel := context.WithCancel(ctx)
+	defer instCancel()
+	servCtx, servCancel := context.WithCancel(context.Background())
+	defer servCancel()
+	servErr := make(chan error, 1)
+	go func() {
+		servErr <- localCtx.serv.Serve(servCtx)
+		instCancel()
+	}()
+	instErr := localCtx.RunInstance(instCtx, 0)
+	servCancel()
+	if err := <-servErr; err != nil {
+		return err
+	}
+	return instErr
 }
 
 func setupLocal(ctx context.Context, cfg *LocalConfig) (*local, context.Context, error) {

